backend/pkg/api/connect/service/user/service.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package user contains the implementation of all User endpoints.
+package user
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"connectrpc.com/connect"
+	"go.uber.org/zap"
+
+	apierrors "github.com/redpanda-data/console/backend/pkg/api/connect/errors"
+	"github.com/redpanda-data/console/backend/pkg/config"
+	"github.com/redpanda-data/console/backend/pkg/console"
+	commonv1alpha1 "github.com/redpanda-data/console/backend/pkg/protogen/redpanda/api/common/v1alpha1"
+	v1alpha1 "github.com/redpanda-data/console/backend/pkg/protogen/redpanda/api/dataplane/v1alpha1"
+	"github.com/redpanda-data/console/backend/pkg/protogen/redpanda/api/dataplane/v1alpha1/dataplanev1alpha1connect"
+	"github.com/redpanda-data/console/backend/pkg/redpanda"
+)
+
+var _ dataplanev1alpha1connect.UserServiceHandler = (*Service)(nil)
+
+// Service that implements the UserServiceHandler interface. This includes all
+// RPCs to manage Redpanda or Kafka users.
+type Service struct {
+	cfg         *config.Config
+	logger      *zap.Logger
+	consoleSvc  console.Servicer
+	redpandaSvc *redpanda.Service
+
+	isProtectedUserFn func(userName string) bool
+}
+
+// NewService creates a new user service handler.
+func NewService(cfg *config.Config,
+	logger *zap.Logger,
+	redpandaSvc *redpanda.Service,
+	consoleSvc console.Servicer,
+	isProtectedUserFn func(userName string) bool,
+) *Service {
+	return &Service{
+		cfg:               cfg,
+		logger:            logger,
+		consoleSvc:        consoleSvc,
+		redpandaSvc:       redpandaSvc,
+		isProtectedUserFn: isProtectedUserFn,
+	}
+}
+
+// ListUsers returns a list of all existing users.
+func (s *Service) ListUsers(ctx context.Context, _ *connect.Request[v1alpha1.ListUsersRequest]) (*connect.Response[v1alpha1.ListUsersResponse], error) {
+	// 1. Check if we can list users
+	if !s.cfg.Redpanda.AdminAPI.Enabled {
+		return nil, apierrors.NewConnectError(
+			connect.CodeUnavailable,
```

Comment:
Should this be `CodeFailedPrecondition` ?

```
// CodeUnavailable indicates that the service is currently unavailable. This
	// is usually temporary, so clients can back off and retry idempotent
	// operations.
```

No amount of backing of or waiting will fix a config issue heh?

Reply:
Indeed it doesn't make sense. I think `UNIMPLEMENTED` would be the most accurate status. The descrition for that status is:

> The operation is not implemented or is not supported/enabled in this service.

I'll change all occurances you found accordingly